src/providers/ibmcloud/gen2.rs

```diff
@@ -0,0 +1,196 @@
+//! Metadata fetcher for IBMCloud (VPC Gen2) instances.
+//!
+//! This provider supports the "VPC Generation 2" infrastructure type
+//! on IBMCloud.
+//! It provides a config-drive as the only metadata source, whose layout
+//! is very similar to `cloud-init NoCloud` [datasource][nocloud], with
+//! a few variations:
+//!  - disk label is `cidata` (lowercase)
+//!  - filesystem is `iso9660`
+//!
+//! nocloud: https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use openssh_keys::PublicKey;
+use slog_scope::{error, warn};
+use tempdir::TempDir;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+const CONFIG_DRIVE_LABEL: &str = "cidata";
+
+/// IBMCloud provider (VPC Gen2).
+#[derive(Debug)]
+pub struct G2Provider {
+    /// Path to the top directory of the mounted config-drive.
+    drive_path: PathBuf,
+    /// Temporary directory for own mountpoint (if any).
+    temp_dir: Option<TempDir>,
+}
+
+impl G2Provider {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        let target =
+            TempDir::new("afterburn").chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            "iso9660",
+            3,
```

Comment:
Aside: for magic number params like this, I like to add a comment beside it like: `// retries`. Or alternatively, name the function something more descriptive so it's obvious from context, e.g. `mount_ro_with_retries`.

Reply:
I share your feeling.

An alternative would be to have its own simple wrapper type, which means in this line we would pass something like `RetriesNum(3)`.

I don't do that by default as more types introduce more cognitive load, but it looks like here you are lamenting the opposite due to ambiguity. A plain newtype or a [shrinkwrap](https://docs.rs/shrinkwraprs/0.2.1/shrinkwraprs/) are the language way to avoid that.

Reply:
Yeah, I essentially had to dig into the codebase here to find out what that param was. Agreed there's a cognitive load involved in trying to embed a solution into the code. Comments go a long way though. See e.g. https://github.com/coreos/rpm-ostree/blob/2589cd1f927c519cd040485c37881e29cf09f0d2/src/app/rpmostree-builtin-upgrade.c#L178.

Reply:
@jlebon so your suggestion, across all the alternatives, is to stick to comments in all consuming place, am I reading it right?

Reply:
I think that'd be better than nothing, yeah. Another near-0 cognitive overload approach is renaming it. Don't want to grow the scope here though. We can circle back to this!